server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryRequestStreamObserver.java

```diff
@@ -0,0 +1,154 @@
+package io.envoyproxy.controlplane.server;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+
+import com.google.protobuf.Any;
+import io.envoyproxy.controlplane.cache.Resources;
+import io.envoyproxy.controlplane.cache.Response;
+import io.envoyproxy.controlplane.cache.Watch;
+import io.envoyproxy.envoy.api.v2.DiscoveryRequest;
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class DiscoveryRequestStreamObserver implements StreamObserver<DiscoveryRequest> {
+  private static final AtomicLongFieldUpdater<DiscoveryRequestStreamObserver> streamNonceUpdater =
```

Comment:
Why not use AtomicLong instead?

Reply:
Please read http://normanmaurer.me/blog/2013/10/28/Lesser-known-concurrent-classes-Part-1/

When creating many instances of classes containing `AtomicLong` a lot of space is wasted:

```
AtomicLong = 24 bytes + 4 bytes (for the reference to it)
volatile long = 8 bytes
This gives us a saving of 16 bytes!
```

Reply:
Ok, that makes sense. Thanks for explaining